edgePy/DGEList.py

```diff
@@ -292,17 +292,40 @@ def rpkm(
         # self = self.cpm(log=log, prior_count=prior_count)
 
     def tpm(
-        self, transcripts: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
-        """Return the DGEList normalized to reads per kilobase of transcript
-        length.
+        self,
+        gene_lengths: np.ndarray,
+        log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+        mean_fragment_lengths: np.ndarray = None,
+    ) -> np.ndarray:
+        """Normalize the DGEList to transcripts per million
+
+        Args:
+            gene_lengths: vector of length for each gene in counts table
+            log: store log outputs
+            prior_count:
+            mean_fragment_lengths: vector of mean fragment length for each sample in counts table (optional)
 
         """
-        raise NotImplementedError
 
-        # TODO: Implement here
+        # based on https://haroldpimentel.wordpress.com/2014/05/08/what-the-fpkm-a-review-rna-seq-expression-units/ and
+        # https://gist.github.com/slowkow/c6ab0348747f86e2748b
+        # how many counts per base
 
-        # self = self.cpm(log=log, prior_count=prior_count)
+        # compute effective length not allowing negative lengths
+        if mean_fragment_lengths:
+            effective_lengths = (gene_lengths[:, None] - mean_fragment_lengths[None, :]).clip(
+                min=1
+            )
+        else:
+            effective_lengths = gene_lengths[:, None]
+
+        base_counts = self.counts / effective_lengths
+
+        self.counts = 10 ** 6 * base_counts / np.sum(base_counts, axis=0)[None, :]
```

Comment:
However you are duplicating code from `DGEList.cpm()`. Please just call that method here instead of copy-paste the four lines. This reduces chances for divergent behavior and simplifies unit testing because if `cpm()` is well unit tested than you can focus more on the unit testing of `tpm`.

Reply:
I agree for the three log_transform lines and will build a private log-function that will be applied by all three normalization methods. Before that I still need to check in the forum that normalization is applied correctly.

however self.counts = ... is different enough (i.e. I would not create an underlying private function that demonstrates that cpm and tpm have one code line relatively in common). 
One